# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1465918523 14400
#      Tue Jun 14 11:35:23 2016 -0400
# Node ID e171851ca3445d7f701b56e4a515744c679f290a
# Parent  67a5cb9cd941f243c19f7049f8c4cecb0dbe561b
hide sys/utsname header

* liboctave/wrappers/uname-wrapper.c,
liboctave/wrappers/uname-wrapper.h: New files.

* liboctave/wrappers/module.mk: Update.

* oct-uname.cc: Use new octave_uname_wrapper function.

diff --git a/liboctave/system/oct-uname.cc b/liboctave/system/oct-uname.cc
--- a/liboctave/system/oct-uname.cc
+++ b/liboctave/system/oct-uname.cc
@@ -22,36 +22,36 @@ along with Octave; see the file COPYING.
 
 #if defined (HAVE_CONFIG_H)
 #  include "config.h"
 #endif
 
 #include <cerrno>
 #include <cstring>
 
-#include <sys/utsname.h>
-
 #include "oct-uname.h"
+#include "uname-wrapper.h"
 
 namespace octave
 {
   namespace sys
   {
     void
     uname::init (void)
     {
-      struct utsname unm;
+      char *sysname, *nodename, *release, *version, *machine;
 
-      err = ::uname (&unm);
+      err = octave_uname_wrapper (&sysname, &nodename, &release,
+                                  &version, &machine);
 
       if (err < 0)
         msg = gnulib::strerror (errno);
       else
         {
-          m_sysname = unm.sysname;
-          m_nodename = unm.nodename;
-          m_release = unm.release;
-          m_version = unm.version;
-          m_machine = unm.machine;
+          m_sysname = sysname;
+          m_nodename = nodename;
+          m_release = release;
+          m_version = version;
+          m_machine = machine;
         }
     }
   }
 }
diff --git a/liboctave/wrappers/module.mk b/liboctave/wrappers/module.mk
--- a/liboctave/wrappers/module.mk
+++ b/liboctave/wrappers/module.mk
@@ -6,32 +6,34 @@ NOINSTALL_WRAPPERS_INC = \
   liboctave/wrappers/hash-wrappers.h \
   liboctave/wrappers/mkostemp-wrapper.h \
   liboctave/wrappers/nanosleep-wrapper.h \
   liboctave/wrappers/nproc-wrapper.h \
   liboctave/wrappers/putenv-wrapper.h \
   liboctave/wrappers/set-program-name-wrapper.h \
   liboctave/wrappers/strftime-wrapper.h \
   liboctave/wrappers/strptime-wrapper.h \
+  liboctave/wrappers/uname-wrapper.h \
   liboctave/wrappers/unsetenv-wrapper.h \
   liboctave/wrappers/vasprintf-wrapper.h
 
 WRAPPERS_SRC = \
   liboctave/wrappers/base64-wrappers.c \
   liboctave/wrappers/canonicalize-file-name-wrapper.c \
   liboctave/wrappers/fpucw-wrapper.c \
   liboctave/wrappers/gen-tempname-wrapper.c \
   liboctave/wrappers/hash-wrappers.c \
   liboctave/wrappers/mkostemp-wrapper.c \
   liboctave/wrappers/nanosleep-wrapper.c \
   liboctave/wrappers/nproc-wrapper.c \
   liboctave/wrappers/putenv-wrapper.c \
   liboctave/wrappers/set-program-name-wrapper.c \
   liboctave/wrappers/strftime-wrapper.c \
   liboctave/wrappers/strptime-wrapper.c \
+  liboctave/wrappers/uname-wrapper.c \
   liboctave/wrappers/unsetenv-wrapper.c \
   liboctave/wrappers/vasprintf-wrapper.c
 
 noinst_LTLIBRARIES += liboctave/wrappers/libwrappers.la
 
 liboctave_wrappers_libwrappers_la_SOURCES = $(WRAPPERS_SRC)
 
 liboctave_wrappers_libwrappers_la_CPPFLAGS = \
diff --git a/liboctave/wrappers/uname-wrapper.c b/liboctave/wrappers/uname-wrapper.c
new file mode 100644
--- /dev/null
+++ b/liboctave/wrappers/uname-wrapper.c
@@ -0,0 +1,63 @@
+/*
+
+Copyright (C) 2016 John W. Eaton
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+// uname may be provided by gnulib.  We don't include gnulib headers
+// directly in Octave's C++ source files to avoid problems that may be
+// caused by the way that gnulib overrides standard library functions.
+
+#if defined (HAVE_CONFIG_H)
+#  include "config.h"
+#endif
+
+#include <sys/utsname.h>
+
+#include "uname-wrapper.h"
+
+// Returns pointers to static storage.
+
+int
+octave_uname_wrapper (char **sysname, char **nodename,
+                      char **release, char **version, char **machine)
+{
+  static struct utsname unm;
+
+  int err = uname (&unm);
+
+  if (err == 0)
+    {
+      *sysname = unm.sysname;
+      *nodename = unm.nodename;
+      *release = unm.release;
+      *version = unm.version;
+      *machine = unm.machine;
+    }
+  else
+    {
+      *sysname = 0;
+      *nodename = 0;
+      *release = 0;
+      *version = 0;
+      *machine = 0;
+    }
+
+  return err;
+}
diff --git a/liboctave/wrappers/uname-wrapper.h b/liboctave/wrappers/uname-wrapper.h
new file mode 100644
--- /dev/null
+++ b/liboctave/wrappers/uname-wrapper.h
@@ -0,0 +1,38 @@
+/*
+
+Copyright (C) 2016 John W. Eaton
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#if ! defined (octave_uname_wrapper_h)
+#define octave_uname_wrapper_h 1
+
+#if defined __cplusplus
+extern "C" {
+#endif
+
+extern int
+octave_uname_wrapper (char **sysname, char **nodename,
+                      char **release, char **version, char **machine);
+
+#if defined __cplusplus
+}
+#endif
+
+#endif
